Documentation/contributing/testing/bpf.rst

```diff
@@ -0,0 +1,307 @@
+marked as failed or skipped.
+
+Sub-tests
+---------
+
+Each ``CHECK`` program may contain sub-tests, each of which have their own test status. A sub-test is
+created with the ``TEST`` macro like so:
+
+.. code-block:: c
+
+    #include "common.h"
+
+    #include <bpf/ctx/xdp.h>
+    #include <lib/jhash.h>
+    #include "bpf/section.h"
+
+    CHECK("xdp", "jhash")
+    int bpf_test(__maybe_unused struct xdp_md *ctx)
+    {
+        test_init();
+
+        TEST("Non-zero", {
+            unsigned int hash = jhash_3words(123, 234, 345, 456);
+
+            if (hash != 2698615579)
+                test_fatal("expected '2698615579' got '%lu'", hash);
+        });
+
+        TEST("Zero", {
+            unsigned int hash = jhash_3words(0, 0, 0, 0);
+
+            if (hash != 459859287)
+                test_fatal("expected '459859287' got '%lu'", hash);
+        });
+
+        test_finish();
+    }
+
+Since all sub-tests are part of the same BPF program they are executed all at once and can share
+the same setup code which can improve run speed and reduce code duplication. Since each tests has
+a name it serves to self-document, and makes it easier to spot what part of a test fails.
+
+Integration tests
+-----------------
+
+Writing tests for a single function or small group of functions should be fairly straightforward, 
+only requiring a ``CHECK`` program. Testing functionality across tail calls requires an additional step,
+Since we will not return to the ``CHECK`` function after making a tail call, we can't check if it was
+successful.
+
+The fix is to use a ``SETUP`` program in addition to a ``CHECK`` program. ``SETUP`` programs with the same
+name will run before the ``CHECK`` program. The context, modified by ``SETUP`` is then passed to ``CHECK``
+which can inspect the result. By executing the test setup and executing the tail call in ``SETUP`` 
+we can execute complete programs. The return code of the ``SETUP`` program is prefixed as ``u32`` to
+the start of the packet data passed to ``CHECK``. 
+
+This is a abbreviated example showing the key components:
+
+.. code-block:: c
+    
+    #include "common.h"
+
+    #include "bpf/ctx/xdp.h"
+    #include "bpf_xdp.c"
+
+    struct {
+        __uint(type, BPF_MAP_TYPE_PROG_ARRAY);
+        __uint(key_size, sizeof(__u32));
+        __uint(max_entries, 2);
+        __array(values, int());
+    } entry_call_map __section(".maps") = {
+        .values = {
+            [0] = &bpf_xdp_entry,
+        },
+    };
+
+    SETUP("xdp", "l2_example")
+    int test1_setup(struct __ctx_buff *ctx)
+    {
+        /* Create room for our packet to be crafted */
+        unsigned int data_len = ctx->data_end - ctx->data;
+        int offset = offset = sizeof(struct ethhdr) - data_len;
+        bpf_xdp_adjust_tail(ctx, offset);
+
+        void *data = (void *)(long)ctx->data;
+        void *data_end = (void *)(long)ctx->data_end;
```

Comment:
Should there be a macro or helper to help doing this?

Reply:
I don't think we have one at the moment. But it does seem like a improvement, would be nice to be able to just set a target packet size instead of having to calculate a relative change. But we will first have to change the code before we can put it in the docs.